certdepot/mongo_depot.go

```diff
@@ -0,0 +1,233 @@
+package certdepot
+
+import (
+	"time"
+
+	"github.com/mongodb/anser/bsonutil"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+	"github.com/pkg/errors"
+	"github.com/square/certstrap/depot"
+	mgo "gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+)
+
+type User struct {
+	ID            string    `bson:"_id"`
+	Cert          string    `bson:"cert"`
+	PrivateKey    string    `bson:"private_key"`
+	CertReq       string    `bson:"cert_req"`
+	CertRevocList string    `bson:"cert_revoc_list"`
+	TTL           time.Time `bson:"ttl"`
+}
+
+var (
+	userIDKey            = bsonutil.MustHaveTag(User{}, "ID")
+	userCertKey          = bsonutil.MustHaveTag(User{}, "Cert")
+	userPrivateKeyKey    = bsonutil.MustHaveTag(User{}, "PrivateKey")
+	userCertReqKey       = bsonutil.MustHaveTag(User{}, "CertReq")
+	userCertRevocListKey = bsonutil.MustHaveTag(User{}, "CertRevocList")
+	userTTLKey           = bsonutil.MustHaveTag(User{}, "TTL")
+)
+
+type mongoCertDepot struct {
+	session        *mgo.Session
+	databaseName   string
+	collectionName string
+	expireAfter    time.Duration
+}
+
+type MgoCertDepotOptions struct {
+	MongoDBURI           string
+	MongoDBDialTimeout   time.Duration
+	MongoDBSocketTimeout time.Duration
+	DatabaseName         string
+	CollectionName       string
+	ExpireAfter          time.Duration
+}
+
+// Create a new cert depot in the specified MongoDB.
+func NewMgoCertDepot(opts MgoCertDepotOptions) (depot.Depot, error) {
+	return newMgoCertDepot(nil, opts)
+}
+
+// Create a new cert depot in the specified MongoDB, using an existing session.
+func NewMgoCertDepotWithSession(s *mgo.Session, opts MgoCertDepotOptions) (depot.Depot, error) {
+	return newMgoCertDepot(s, opts)
+}
+
+func newMgoCertDepot(s *mgo.Session, opts MgoCertDepotOptions) (depot.Depot, error) {
+	opts.validate()
+
+	if s == nil {
+		var err error
+		s, err = mgo.DialWithTimeout(opts.MongoDBURI, opts.MongoDBDialTimeout)
+		if err != nil {
+			return nil, errors.Wrapf(err, "could not connect to db %s", opts.MongoDBURI)
+		}
+		s.SetSocketTimeout(opts.MongoDBSocketTimeout)
+	}
+
+	return &mongoCertDepot{
+		session:        s,
+		databaseName:   opts.DatabaseName,
+		collectionName: opts.CollectionName,
+		expireAfter:    opts.ExpireAfter,
+	}, nil
+}
+
+func (opts *MgoCertDepotOptions) validate() {
```

Comment:
error?

Reply:
it is always nil since we set defaults for everything, should we just return nil?